src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +105,101 @@
+
+(schema/defn cleanup-and-refill-pool
+  "Cleans up the given instances and optionally refills the pool with
+  new instances. Should only be called from the modify-instance-agent"
   [pool-context :- jruby-schemas/PoolContext
-   old-pool-state :- jruby-schemas/PoolState
-   new-pool-state :- jruby-schemas/PoolState
+   pool-state :- jruby-schemas/PoolState
+   old-instances :- [JRubyInstance]
    refill? :- schema/Bool]
-  (let [{:keys [config]} pool-context
-        pool-state-atom (jruby-internal/get-pool-state-container pool-context)
-        new-pool (:pool new-pool-state)
-        old-pool (:pool old-pool-state)
-        old-pool-size (:size old-pool-state)
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)
+        new-instance-ids (map inc (range pool-size))
+        config (:config pool-context)
         cleanup-fn (get-in config [:lifecycle :cleanup])]
-    (log/info "Replacing old JRuby pool with new instance.")
-    (reset! pool-state-atom new-pool-state)
-    (log/info "Swapped JRuby pools, beginning cleanup of old pool.")
-    (doseq [i (range old-pool-size)]
+    (doseq [[old-instance new-id] (zipmap old-instances new-instance-ids)]
       (try
-        (let [id (inc i)
-              instance (jruby-internal/borrow-from-pool!*
-                        jruby-internal/borrow-without-timeout-fn
-                        old-pool)]
-          (try
-            (jruby-internal/cleanup-pool-instance! instance cleanup-fn)
-            (when refill?
-              (jruby-internal/create-pool-instance! new-pool id config
-                                                    (partial send-flush-instance! pool-context))
-              (log/infof "Finished creating JRubyInstance %d of %d"
-                         id old-pool-size))
-            (finally
-              (.releaseItem old-pool instance false))))
+        (jruby-internal/cleanup-pool-instance! old-instance cleanup-fn)
+        (when refill?
+          (jruby-internal/create-pool-instance! pool new-id config
+                                                (partial send-flush-instance! pool-context))
+          (log/infof "Finished creating JRubyInstance %d of %d"
+                     new-id pool-size))
         (catch Exception e
-          (.clear new-pool)
-          (.insertPill new-pool (PoisonPill. e))
+          (.clear pool)
+          (send-and-await-fill-with-poison-pills! pool-context
+                                                  (partial jruby-internal/insert-poison-pill e))
           (throw (IllegalStateException.
-                  "There was a problem adding a JRubyInstance to the pool."
-                  e)))))
-    ;; Add a "RetryPoisonPill" to the pool in case something else is in the
-    ;; process of borrowing from the old pool.
-    (.insertPill old-pool (RetryPoisonPill. old-pool))))
+                  "There was a problem creating a JRubyInstance for the pool."
+                  e))))))
+  (log/info "Finished draining pool."))
```

Comment:
Maybe we should make the draining/refilling message conditional here based on what `refill?` is set to?